File: client/python/modeldb/events/FitEvent.py
"""
Event indicating estimator was used to fit model.
"""
from modeldb.events.Event import Event
from ..thrift.modeldb import ttypes as modeldb_types
import json


class FitEvent(Event):
    """
    Class for creating and storing FitEvents
    """

    def __init__(self, model, spec, df, metadata={}):
        self.model = model
        self.spec = spec
        self.df = df
        self.metadata = metadata

    def make_event(self, syncer):
        """
        Constructs a thrift FitEvent object with appropriate fields.
        """
        syncable_transformer = syncer.convert_model_to_thrift(self.model)
        model_spec = syncer.convert_spec_to_thrift(self.spec)
        syncable_dataframe = syncer.convert_df_to_thrift(self.df)
        columns = syncer.set_columns(self.df)
        fe = modeldb_types.FitEvent(syncable_dataframe, model_spec,
                                    syncable_transformer, columns,
                                    [], [], syncer.experiment_run.id)
        fe.metadata = json.dumps(self.metadata, default=str)
        return fe

    def associate(self, res, syncer):
        """
        Stores the generated ids into dictionary.
        """
        syncer.store_object(self.df, res.dfId)
        syncer.store_object(self.spec, res.specId)
        syncer.store_object(self.model, res.modelId)
        syncer.store_object(self, res.eventId)

    def sync(self, syncer):
        """
        Stores FitEvent on the server.
        """
        fe = self.make_event(syncer)
        thrift_client = syncer.client
        res = thrift_client.storeFitEvent(fe)
        self.associate(res, syncer)

File: client/scala/libs/spark.ml/src/main/scala-2.11/edu/mit/csail/db/ml/modeldb/client/event/FitEvent.scala
package edu.mit.csail.db.ml.modeldb.client.event

import com.twitter.util.Await
import edu.mit.csail.db.ml.modeldb.client._
import modeldb.ModelDBService.FutureIface
import org.apache.spark.ml.{PipelineStage, SyncableEstimator, Transformer}
import org.apache.spark.sql.DataFrame

/**
  * Event indicating that an Estimator was used to fit a model.
  *
  * @param estimator - The estimator performing the fitting.
  *                  It's a pipeline stage because we can't easily
  *                  type parameterize all Estimators.
  * @param dataframe - The data being fit.
  * @param model - The model produced by the fitting procedure.
  */
case class FitEvent(estimator: PipelineStage,
                    dataframe: DataFrame,
                    model: Transformer) extends ModelDbEvent {

  def makeEvent(mdbs: ModelDbSyncer) = {
    modeldb.FitEvent(
      SyncableDataFrame(dataframe),
      SyncableEstimator(estimator),
      SyncableTransformer(model),
      mdbs.featureTracker.getFeatureCols(dataframe, model),
      mdbs.featureTracker.getOutputCols(model),
      mdbs.featureTracker.getLabelColumns(model),
      experimentRunId = mdbs.experimentRun.id,
      problemType = SyncableProblemType(model)
    )
  }

  def associate(fer: modeldb.FitEventResponse, mdbs: ModelDbSyncer) = {
    mdbs.associateObjectAndId(dataframe, fer.dfId)
      .associateObjectAndId(estimator, fer.specId)
      .associateObjectAndId(model, fer.modelId)
      .associateObjectAndId(this, fer.eventId)
  }
  override def sync(client: FutureIface, mdbs: Option[ModelDbSyncer]): Unit = {
    val res = Await.result(client.storeFitEvent(makeEvent(mdbs.get)))
    SyncableSpecificModel(res.modelId, model, Some(client), mdbs.get.shouldStoreSpecificModels)
    associate(res, mdbs.get)
  }
}